File: app/src/main/java/com/shumidub/todoapprealm/ui/fragment/note_fragment/FolderNoteFragment.java
package com.shumidub.todoapprealm.ui.fragment.note_fragment;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Bundle;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import androidx.appcompat.app.ActionBar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.ItemTouchHelper;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.LinearLayout;

import com.shumidub.todoapprealm.App;
import com.shumidub.todoapprealm.R;
import com.shumidub.todoapprealm.realmcontrollers.notescontroller.FolderNotesRealmController;
import com.shumidub.todoapprealm.realmmodel.notes.FolderNotesObject;
import com.shumidub.todoapprealm.ui.actionmode.EmptyActionModeCallback;
import com.shumidub.todoapprealm.ui.actionmode.note.FolderNoteActionModeCallback;
import com.shumidub.todoapprealm.ui.activity.main.MainActivity;
import com.shumidub.todoapprealm.ui.dialog.note_dialog.AddNoteDialog;
import com.shumidub.todoapprealm.ui.dialog.note_dialog.EditNoteDialog;

import io.realm.RealmList;

import static com.shumidub.todoapprealm.ui.dialog.note_dialog.AddNoteDialog.TYPE_NOTE;


/**
 * Created by Артем on 19.12.2017.
 *
 */


public class FolderNoteFragment extends Fragment{

    ActionBar actionBar;
    public boolean actionModeIsEnabled = false;

    RecyclerView rv;
    LinearLayout emptyState;

    FolderNotesRecyclerViewAdapter folderAdapter;
    NotesRecyclerViewAdapter noteAdapter;

    int type = AddNoteDialog.TYPE_FOLDER;

    public boolean isNoteFragment = false;

    String title = "Notes";

    long idFolderNoteObject = 0;
    long idNoteObject = 0;
    long id;


    boolean folderViewShowing= true;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.note_fragment_layout, container, false);
        actionBar = ((MainActivity) getActivity()).getSupportActionBar();
        return view;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        emptyState =  view.findViewById(R.id.empty_state);
        rv = view.findViewById(R.id.recycle_view);
        rv.setLayoutManager(new LinearLayoutManager(getContext()));
        setFolderNoteViews();
        actionBar = ((MainActivity) getActivity()).getSupportActionBar();
        setHasOptionsMenu(true);
        setTouchHelper(rv);
        id = idFolderNoteObject;


    }



    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);

        MenuItem add = menu.add(2,2,2,"add ");
        add.setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS);
        add.setIcon(R.drawable.ic_add);
        add.setOnMenuItemClickListener((MenuItem a) -> {
            AddNoteDialog addNoteDialog = AddNoteDialog.newInstance(type, id);
            try {
                addNoteDialog.show(getActivity().getSupportFragmentManager(), "Add_note");
                InputMethodManager imm = (InputMethodManager) getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
                imm.toggleSoftInputFromWindow(
                        getActivity().getWindow().getDecorView().getApplicationWindowToken(),
                        InputMethodManager.SHOW_FORCED, 0);
            }catch (NullPointerException e){e.printStackTrace();}


//            new AddReportDialog().show(getActivity().getSupportFragmentManager(), AddReportDialog.ADD_REPORT_TITLE);
            return true;
        });

    }

    public void notifyDataChanged(){

        RecyclerView.Adapter adapter = rv.getAdapter();

        if (adapter instanceof FolderNotesRecyclerViewAdapter){
            folderAdapter.notifyDataSetChanged();
        } else{
            noteAdapter.notifyDataSetChanged();
        }

        if (adapter.getItemCount() == 0){
            emptyState.setVisibility(View.VISIBLE);
        } else {
            emptyState.setVisibility(View.GONE);
        }
    }



    @SuppressLint("RestrictedApi")
    public void setFolderNoteViews(){


        actionBar.setDisplayHomeAsUpEnabled(false);

        title = "Notes";
        if ( ((MainActivity)getActivity()).getPagerAdapterPosition() == 0 )
            actionBar.setTitle(title);


        type = AddNoteDialog.TYPE_FOLDER;
        isNoteFragment = false;

        folderAdapter = new FolderNotesRecyclerViewAdapter((MainActivity) getActivity());
        folderAdapter.setOnClickListener((h,p,idFolderFromAdapter)->{
            setNoteViews(idFolderFromAdapter);
        });
        folderAdapter.setOnLongClickListener((h,p,id1)->{
            actionBar.startActionMode(new FolderNoteActionModeCallback()
                            .getFolderNoteActionMode((MainActivity) getActivity(), this,
                                    EditNoteDialog.TYPE_FOLDER, id1));

            return true;
        });
        rv.setAdapter(folderAdapter);

        if (folderAdapter.getItemCount() == 0){
            emptyState.setVisibility(View.VISIBLE);
        } else {
            emptyState.setVisibility(View.GONE);
        }

    }

    @SuppressLint("RestrictedApi")
    public void setNoteViews(long idFolderFromAdapter){


        
        actionBar.setDisplayHomeAsUpEnabled(true);
        
        title = FolderNotesRealmController.getFolderNote(idFolderFromAdapter).getName();
        actionBar.setTitle(title);

        isNoteFragment = true;

        type = TYPE_NOTE;
        id = idFolderFromAdapter;

        noteAdapter = new NotesRecyclerViewAdapter(idFolderFromAdapter, (MainActivity) getActivity());
        noteAdapter.setId(idFolderFromAdapter);
        rv.setAdapter(noteAdapter);
        noteAdapter.setOnClickListener((h,p,id)-> setFolderNoteViews());
        noteAdapter.setOnLongClickListener((h,p,id)->{
            actionBar.startActionMode(new FolderNoteActionModeCallback()
                    .getFolderNoteActionMode((MainActivity) getActivity(), this,
                            EditNoteDialog.TYPE_NOTE, id));
            return true;
        });
        rv.setAdapter(noteAdapter);

        if (noteAdapter.getItemCount() == 0){
            emptyState.setVisibility(View.VISIBLE);
        } else {
            emptyState.setVisibility(View.GONE);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if(item.getItemId() == android.R.id.home ){
            setFolderNoteViews();
            actionBar.setTitle(title);
        }
        return super.onOptionsItemSelected(item);
    }

    @SuppressLint("RestrictedApi")
    public void finishActionMode(){
        actionBar.startActionMode(new EmptyActionModeCallback());
    }


    private void setTouchHelper(RecyclerView rv){


        // set ITEM TOUCH HELPER for folder rv
        App.initRealm();
        RealmList<FolderNotesObject> folderOfNotesContainerList = App.folderOfNotesContainerList;
        //todo try is it working, or need not use linked variable

        ItemTouchHelper itemTouchHelper = new ItemTouchHelper(new ItemTouchHelper.SimpleCallback(
                ItemTouchHelper.UP | ItemTouchHelper.DOWN ,0) {

            int dragFrom = -1;
            int dragTo = -1;

            @SuppressLint("RestrictedApi")
            @Override
            public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder,
                                  RecyclerView.ViewHolder target) {

                    actionBar.startActionMode(new EmptyActionModeCallback());

                int fromPosition = viewHolder.getAdapterPosition();
                int toPosition = target.getAdapterPosition();

                if (dragFrom == -1) {
                    dragFrom = fromPosition;
                }
                dragTo = toPosition;

                rv.getAdapter().notifyItemMoved(fromPosition, toPosition);
                return true;
            }

            // todo need fix if move bellow "add list"
            private void reallyMoved(int from, int to) {

                RecyclerView.Adapter adapter = rv.getAdapter();

                if (adapter instanceof FolderNotesRecyclerViewAdapter){
                    ((FolderNotesRecyclerViewAdapter) adapter).itemMove(from, to);
                }else if (adapter instanceof NotesRecyclerViewAdapter){
                    ((NotesRecyclerViewAdapter) adapter).itemMove(from, to);
                }

            }

            @Override
            public void clearView(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
                super.clearView(recyclerView, viewHolder);

                if (dragFrom != -1 && dragTo != -1 && dragFrom != dragTo) {
                    reallyMoved(dragFrom, dragTo);
                }
                dragFrom = dragTo = -1;
            }

            @Override
            public void onSwiped(final RecyclerView.ViewHolder viewHolder, int swipeDir) { }

        });

        itemTouchHelper.attachToRecyclerView(rv);


    }


    public String getValidTitle(){
       return title;
    }



}

File: app/src/main/java/com/shumidub/todoapprealm/ui/fragment/note_fragment/FolderNoteFragment.kt
package com.shumidub.todoapprealm.ui.fragment.note_fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.platform.ViewCompositionStrategy
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.shumidub.todoapprealm.R
import com.shumidub.todoapprealm.ui.notes.NotesDialog
import com.shumidub.todoapprealm.ui.notes.NotesMode
import com.shumidub.todoapprealm.ui.notes.NotesScreen
import com.shumidub.todoapprealm.ui.notes.NotesViewModel
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch

@AndroidEntryPoint
class FolderNoteFragment : Fragment() {

    private val viewModel: NotesViewModel by viewModels()

    // Mirrored from ViewModel state for MainActivity field access via Java.
    @JvmField var actionModeIsEnabled: Boolean = false
    @JvmField var isNoteFragment: Boolean = false

    private var currentTitle: String = DEFAULT_TITLE

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        setHasOptionsMenu(true)
        return ComposeView(requireContext()).apply {
            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
            setContent { NotesScreen(viewModel) }
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewLifecycleOwner.lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.state.collect { state ->
                    syncWithLegacyActionBar(state.mode, state.dialog)
                }
            }
        }
    }

    private fun syncWithLegacyActionBar(mode: NotesMode, dialog: NotesDialog) {
        isNoteFragment = mode is NotesMode.Notes
        actionModeIsEnabled = dialog is NotesDialog.FolderActions ||
            dialog is NotesDialog.NoteActions

        currentTitle = when (mode) {
            NotesMode.Folders -> DEFAULT_TITLE
            is NotesMode.Notes -> mode.folderName.ifBlank { DEFAULT_TITLE }
        }

        val actionBar = (activity as? AppCompatActivity)?.supportActionBar ?: return
        actionBar.title = currentTitle
        actionBar.setDisplayHomeAsUpEnabled(isNoteFragment)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        menu.add(2, 2, 2, "add")
            .setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS)
        menu.findItem(2).setIcon(R.drawable.ic_add)
        menu.findItem(2).setOnMenuItemClickListener {
            viewModel.openAdd()
            true
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            viewModel.backToFolders()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    // Compatibility shims for MainActivity.onBackPressed
    fun finishActionMode() {
        viewModel.dismissDialog()
        actionModeIsEnabled = false
    }

    fun setFolderNoteViews() {
        viewModel.backToFolders()
    }

    fun getValidTitle(): String = currentTitle

    private companion object {
        const val DEFAULT_TITLE = "Notes"
    }
}
